### PR TITLE
 DOC: updated documentation for BusinessHour and BusinessDay

### PR DESCRIPTION
**Pull request by natmokval**
This PR is related to PR #50182.
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

Updated documentation for `BusinessDay`, `BusinessHour`, `CustomBusinessHour`. Corrected lists of parameters and provided more examples.
Could you, please, @MarcoGorelli, review my PR?

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -1610,29 +1628,38 @@ cdef class BusinessHour(BusinessMixin):
     Parameters
     ----------
     n : int, default 1
-        The number of months represented.
+        The number of hours represented.
     normalize : bool, default False
         Normalize start/end dates to midnight before generating date range.
-    weekmask : str, Default 'Mon Tue Wed Thu Fri'
-        Weekmask of valid business days, passed to ``numpy.busdaycalendar``.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1617]: I guess a bit of copypasta went on here 😄  nice catch

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -1610,29 +1628,38 @@ cdef class BusinessHour(BusinessMixin):
...
     start : str, time, or list of str/time, default "09:00"
         Start time of your custom business hour in 24h format.
     end : str, time, or list of str/time, default: "17:00"
         End time of your custom business hour in 24h format.
 
     Examples
     --------
-    >>> from datetime import time
+    You can use the parameter ``n`` to represent the shift to n hours.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1641]: perhaps "to represent a shift of n hours" would be clearer? likewise for above

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -3662,7 +3690,7 @@ cdef class CustomBusinessHour(BusinessHour):
     >>> ts + pd.offsets.CustomBusinessHour(end=dt_time(19, 0))
     Timestamp('2022-08-08 10:00:00')
 
-    In the example below we divide our business day hours into several parts.
+    You can divide your business day hours into several parts.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:3693]: thanks for getting back to this!

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -1610,29 +1628,38 @@ cdef class BusinessHour(BusinessMixin):
...
-    >>> ts + pd.offsets.BusinessHour(start=[time(9, 0), "20:00"],
-    ...                              end=["17:00", time(22, 0)])
-    Timestamp('2022-08-05 20:00:00')
+
+    >>> from datetime import time as dt_time
+    >>> ts = pd.Timestamp(2022, 8, 5, 22)
+    >>> ts + pd.offsets.BusinessHour(end=dt_time(19, 0))
+    Timestamp('2022-08-08 10:00:00')
+
+    The parameter ``normalize`` equal True forces shift to midnight.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1658]: I think it's clearer how you wrote it above

    Passing the parameter ``normalize`` equal True, you shift the start
    of next business hour to midnight.

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -1610,29 +1628,38 @@ cdef class BusinessHour(BusinessMixin):
...
+
     >>> ts = pd.Timestamp(2022, 8, 5, 16)
-    >>> ts + pd.offsets.BusinessHour()
-    Timestamp('2022-08-08 09:00:00')
     >>> ts + pd.offsets.BusinessHour(start="11:00")
     Timestamp('2022-08-08 11:00:00')
-    >>> ts + pd.offsets.BusinessHour(end=time(19, 0))
-    Timestamp('2022-08-05 17:00:00')
-    >>> ts + pd.offsets.BusinessHour(start=[time(9, 0), "20:00"],
-    ...                              end=["17:00", time(22, 0)])
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1634]: do we want to lose the example of where `start` and `end` are lists? I think it'd be OK to copy-and-paste from the example you added to CustomBusinessHour - repetition in documentation is generally OK (as opposed to in code where we generally don't want to repeat ourselves too much)

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks for working on this @natmokval ! Got some minor comments, but overall this is a nice improvement

**Comment by natmokval** [comment]
> Thanks for working on this @natmokval ! Got some minor comments, but overall this is a nice improvement

Thank you @MarcoGorelli, I like doing it. I find the idea of slightly improving the documentation of offsets reasonable.

Agreed, I did copy-and-paste this example from CustomBusinessHour. The PR is updated as you suggested.

File: pandas/_libs/tslibs/offsets.pyx [hunk 6]
```diff
@@ -1655,11 +1655,26 @@ cdef class BusinessHour(BusinessMixin):
...
+    of next business hour to midnight.
 
     >>> ts = pd.Timestamp(2022, 12, 9, 8)
     >>> ts + pd.offsets.BusinessHour(normalize=True)
     Timestamp('2022-12-09 00:00:00')
+
+    You can divide your business day hours into several parts.
+
+    >>> import datetime as dt
+    >>> freq = pd.offsets.CustomBusinessHour(start=["06:00", "10:00", "15:00"],
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1668]: I guess this should still be `BusinessHour`?

Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: Oops! I corrected my mistake. The copy-and-paste method can be dangerous sometimes.

File: pandas/_libs/tslibs/offsets.pyx [hunk 7]
```diff
@@ -1494,11 +1494,29 @@ cdef class BusinessDay(BusinessMixin):
...
+        The number of days represented.
+    normalize : bool, default False
+        Normalize start/end dates to midnight before generating date range.
+
     Examples
     --------
-    >>> ts = pd.Timestamp(2022, 8, 5)
-    >>> ts + pd.offsets.BusinessDay()
-    Timestamp('2022-08-08 00:00:00')
+    You can use the parameter ``n`` to shift the next business day to n days.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1506]: I think the way you wrote it in the other one is clearer:

    You can use the parameter ``n`` to represent a shift of n business days.

File: pandas/_libs/tslibs/offsets.pyx [hunk 8]
```diff
@@ -1494,11 +1494,29 @@ cdef class BusinessDay(BusinessMixin):
     """
     DateOffset subclass representing possibly n business days.
 
+    Parameters
+    ----------
+    n : int, default 1
+        The number of days represented.
+    normalize : bool, default False
+        Normalize start/end dates to midnight before generating date range.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1502]: Can this be used outside of generating a date range? If so, maybe

    Normalize start/end dates to midnight

is enough?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Nice - CI failures are unrelated and should be resolved now, I've just made some final suggestions - if you address them and add a commit then CI should hopefully be green

File: pandas/_libs/tslibs/offsets.pyx (hunk 8 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: Yes, I think so. Changed this line.

Thank you for your comments.  I created a new commit.

File: pandas/_libs/tslibs/offsets.pyx [hunk 9]
```diff
@@ -1494,11 +1494,29 @@ cdef class BusinessDay(BusinessMixin):
...
+    You can use the parameter ``n`` to represent a shift of n business days.
+
+    >>> ts = pd.Timestamp(2022, 12, 9, 15)
+    >>> ts.strftime('%a %d %b %Y %H:%M')
+    'Fri 09 Dec 2022 15:00'
+    >>> (ts + pd.offsets.BusinessDay(n=5)).strftime('%a %d %b %Y %H:%M')
+    'Fri 16 Dec 2022 15:00'
+
+    Passing the parameter ``normalize`` equal True, you shift the start
+    of next business day to midnight.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1515]: equal True -> equal to True

of next business day -> of the next business day

File: pandas/_libs/tslibs/offsets.pyx [hunk 10]
```diff
@@ -1610,29 +1628,53 @@ cdef class BusinessHour(BusinessMixin):
...
-    ...                              end=["17:00", time(22, 0)])
-    Timestamp('2022-08-05 20:00:00')
+
+    >>> from datetime import time as dt_time
+    >>> ts = pd.Timestamp(2022, 8, 5, 22)
+    >>> ts + pd.offsets.BusinessHour(end=dt_time(19, 0))
+    Timestamp('2022-08-08 10:00:00')
+
+    Passing the parameter ``normalize`` equal True, you shift the start
+    of next business hour to midnight.
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/offsets.pyx:1659]: same comment as above

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Almost there 💪 

There's some CI failures from the docstring validation script, check https://github.com/pandas-dev/pandas/actions/runs/3707564825/jobs/6284132896

**Comment by natmokval** [comment]
> Almost there 💪
> 
> There's some CI failures from the docstring validation script, check https://github.com/pandas-dev/pandas/actions/runs/3707564825/jobs/6284132896

Thank you MarcoGorelli, for your help. I checked the errors raised by the docstring validation script in CI and fixed them. It's interesting, why if I run the docstring validation script locally these errors don't occur.

File: pandas/_libs/tslibs/offsets.pyx (hunk 9 above)
Comment (natmokval) [reply pandas/_libs/tslibs/offsets.pyx]: I corrected my grammar mistakes and updated my PR.

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, nice improvement, thanks @natmokval !

**Comment by MarcoGorelli** [comment]
> I checked the errors raised by the docstring validation script in CI and fixed them. It's interesting, why if I run the docstring validation script locally these errors don't occur.

My guess is that you hadn't re-compiled your C extensions (`python setup.py build_ext`)

**Comment by natmokval** [comment]
Thank you MarcoGorelli, for your comment. Usually, I run `python setup.py build_ext -j 1 `, `python -m pip install -e . --no-build-isolation --no-use-pep517`and smth like `python scripts/validate_docstrings.py pandas.offsets.BusinessDay` after modifying  .pyx files. Otherwise, html won’ be rebuilt.
Probably this time I missed something.

On the other hand, when I tried to reproduce my mistake, I run `python setup.py build_ext -j 1 --force`, and then 
the docstring validation script raised my error. Maybe it would be better to use `--force` while re-compiling the C extensions.

**Comment by MarcoGorelli** [comment]
if you just do `python setup.py build_ext`, then only the files that have changed since you last compiled will be recompiled, whereas `--force` recompiles everything - usually the first one's fine (it's what I usually do), but occasionally things go wrong and I have to use `--force`. hopefully the meson build will make this work a lot better - there's a PR open for that, don't know how far away it is from being ready though

**Pull request merged**